enhancements/catalog-data-transfer-reduction.md

```diff
@@ -0,0 +1,463 @@
+
+#### Produce a catalog of all channel’s heads from a starting catalog
+
+I should be able to prune down a starting catalog containing one or more
+package channels to only their head bundles. Optionally, dependencies
+of these heads are included. Including heads in catalog output should
+be a mode of other types of pruning/diffing.
+
+#### Provide catalog manipulation functionality as both a command and library
+
+I want to have the option to import functionality described here as a library
+to wrap in some binary, or as a CLI tool that calls the library (`opm`).
+
+#### Define the on disk representation of a differential data set (catalog)
+
+I should be able to configure pruning and diffing catalogs in the same manner,
+and importantly, in a single step, and have an output commonly understood by catalog manipulation tools.
+
+#### Support a simple and controllable approach to updating the disconnected side’s catalog
+
+Once a diff is produced and mirrored, I should be able to add or merge the diff
+with an existing catalog seamlessly. This addition/merge operation should be verified
+to work on the connected side prior to mirroring the diff.
+
+### Non-Goals
+
+1. Perform the mirroring itself.
+1. Resolve dependencies using OLM’s dep resolver on the connected side.
+    - This is a candidate for follow-up work.
+1. Decide which architectures to mirror.
+
+## Proposal
+
+### User Stories [optional]
+
+<!-- TODO maybe -->
+
+### Implementation Details/Notes/Constraints [optional]
+
+This section delves into [declarative configs][declcfg-ep] (DC) heavily so it is recommended you read that EP first.
+
+Some terminology:
+- *Bundle*: a collection of manifests that define an operator at a particular state (ex. version).
+- *Replaces*: a [reference][replaces] to another bundle that can be upgraded from to the referent.
+- *Skips*: a [reference][skips] to another bundle that can be skipped when upgrading to the referent.
+- *Dependency*: a package or group-version-kind (GVK) requirement by a bundle.
+  - There are three ways to specify dependencies*: by full package, by package version range, and by GVK.
+  The first way is effectively the same as the second, with version range being the entire set of published versions.
+- *Upgrade graph*: the DAG of replaces from one bundle to another, skipping bundles referenced in skips.
+- *Replaces chain*: upgrade graph, ignoring skips.
+- *Channel*: a named upgrade graph.
+- *Channel head*: a bundle in a channel that is not referenced by any replaces or skips.
+- *Package*: a collection of channels.
+- *Index*: a collection of packages, each of which may or may not contain all published bundles for that package.
+  - Typically built into an image that can serve packages via gRPC when run to OLM via a CatalogSource
+- *Catalog*: one or more indices serving content.
+
+Catalog invariants:
+- Bundles, channels, and packages have unique names within their parent type (bundle->channel->package->catalog).
```

Comment:
There's another invariant which is that the csv name has to be unique in the entire catalog.

Reply:
(but we should definitely not use csv name as an identifier, as we want to create bundle formats that do not have csvs at all)

Reply:
Is the ultimate invariant that a bundle will have a unique name within a catalog, or package name + bundle name will be unique?